addons/survey/models/survey_survey.py

```diff
@@ -430,63 +438,63 @@ def _get_number_of_attempts_lefts(self, partner, email, invite_token):
-
-        current_page_index = pages_or_questions.index(next(p for p in pages_or_questions if p[1].id == page_or_question_id))
-        # is first page
-        if current_page_index == 0:
-            return None
-
-        previous_page_id = pages_or_questions[current_page_index - 1][1].id
+            return self.question_ids if self.questions_layout == 'page_per_question' else self.page_ids
 
-        return previous_page_id
-
-    @api.model
-    def next_page_or_question(self, user_input, page_or_question_id):
-        """ The next page to display to the user, knowing that page_id is the id
-            of the last displayed page.
-
-            If page_id == 0, it will always return the first page of the survey.
-
-            If all the pages have been displayed, it will return None
-
-            .. note::
-                It is assumed here that a careful user will not try to set go_back
-                to True if she knows that the page to display is the first one!
-                (doing this will probably cause a giant worm to eat her house)
+    def _get_next_question(self, user_input, page_or_question_id, go_back=False):
+        """
+        This methods gets the next question or page to display.
+        In case the survey contains conditional question, when navigating to next or previous questions,
+        in page_per_question layout only, the next or previous question to display depends on the selected answers
+        and the questions hierarchy. This methods returns the next question index to display among the survey_questions.
+        :param user_input: user's answers
+        :param page_or_question_id: current page or question id
+        :param go_back: must be True to get the previous question
+        :return: next or previous question.id
         """
 
         survey = user_input.survey_id
-
         pages_or_questions = survey._get_pages_or_questions(user_input)
-        if not pages_or_questions:
-            return (None, False)
+        Question = self.env['survey.question']
+
+        # Get Next
+        if not go_back:
+            if not pages_or_questions:
+                return Question
+            # First page
+            if page_or_question_id == 0:
+                return pages_or_questions[0]
+
+        current_page_index = pages_or_questions.ids.index(next(p.id for p in pages_or_questions if p.id == page_or_question_id))
+
+        # Get previous and we are on first page  OR Get Next and we are on last page
+        if (go_back and current_page_index == 0) or current_page_index == len(pages_or_questions) - 1:
+            return Question
+
+        # Conditional Questions Management
+        triggering_answer_by_question, triggered_questions_by_answer, selected_answers = user_input._get_conditional_values()
+        if survey.has_conditional_questions and survey.questions_layout == 'page_per_question' and triggered_questions_by_answer:
+            while True:
```

Comment:
Why not
  * create a list of next_pages_or_questions (order depending on go_back)
  * iterate on it, each iteration
    * pop first item
    * check conditional condition if any
    * break or next loop

That way we avoid manual index computation with +1 / -1 depending on go_back, and we do'nt have a while Trye ... and it is easier to understand, aka "take the first one backward or forward tha meets the conditions"

What do you think ?